In [1]:
import pandas as pd
from sklearn import datasets   # to import the dataset of boston present in sklearn 
from sklearn.model_selection import train_test_split # to split the train and test dataset
from sklearn.linear_model import LinearRegression # to find the linear regression
from sklearn.metrics import mean_squared_error #to find the error in prediction


In [2]:
boston = datasets.load_boston() # importing the boston dataset from the scikitlearn

In [3]:
print(boston.DESCR) #describes the dataset


.. _boston_dataset:

Boston house prices dataset
---------------------------

**Data Set Characteristics:**  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pu

value of X=13,  y = medv ,
number of observation=n=506 ,

X=feature columns or independent variable ,
y=target or dependent variable


In [4]:
X=pd.DataFrame(boston.data,columns=boston.feature_names)
y=pd.DataFrame(boston.target,columns=['MEDV'])

X #prints the FEATURE values of X 

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4    0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786  1.0  273.0   
502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875  1.0  273.0   
503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675  1.0  273.0   
504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889  1.0  273.0   
505  0.04741   0.0  11.93   0.0  0.573  6.030  80.8  2.5050  1.0  273.0   

     PTRATIO       B  LSTAT  
0       15.3  396.90   4.98  
1       17.8  396.90   9.14  
2       17.8  392.83   4.03  
3       18.7  394.63   2.94  
4       18.7  396.90   5.33  
..       ...     ...    ...  
501     21.0  391.99   9.67  
502     21.0  396.90   9.08  
503     21.0  396.90   5.64  
504     21.0  393.45   6.48  
505     21.0  396.90   7.88  

[506 rows x 13 columns]

In [5]:
y # PRINTS THE TARGET VALUE 

MEDV
0    24.0
1    21.6
2    34.7
3    33.4
4    36.2
..    ...
501  22.4
502  20.6
503  23.9
504  22.0
505  11.9

[506 rows x 1 columns]

In [6]:
X.isnull().sum() # used for checking for null values in X dataset

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
dtype: int64

In [7]:
y.isnull().sum() # used to check for null values in y dataset

MEDV    0
dtype: int64

In [38]:
# used to split the data set into train,test pair
# train =70%,test=30%

train_x,test_x,train_y,test_y=train_test_split(X,y,test_size=0.3,random_state=1)


In [39]:
# FIND THE REGRESSION MODEL OFTHE GIVEN TRAIN DATA
reg_model=LinearRegression().fit(train_x,train_y)

In [40]:
# predicting the target value using the test dataset
pred_y= reg_model.predict(test_x)

In [41]:
pred_y

array([[32.3914647 ],
       [27.94401279],
       [17.83762764],
       [21.66941388],
       [18.93639614],
       [20.01995795],
       [32.63414961],
       [17.86179976],
       [24.73211676],
       [26.77346826],
       [27.26138023],
       [28.63243014],
       [21.16103038],
       [27.15098589],
       [23.33198614],
       [21.02894005],
       [17.11819755],
       [37.52455938],
       [30.69864427],
       [ 8.46104486],
       [20.83966032],
       [16.26732451],
       [25.05030307],
       [24.8069032 ],
       [31.6789299 ],
       [10.98989363],
       [13.54111562],
       [16.54697099],
       [35.86699683],
       [14.69589266],
       [21.05655107],
       [13.90301353],
       [43.01321262],
       [17.80335305],
       [21.68817023],
       [20.45077825],
       [17.38095983],
       [26.81334904],
       [10.08706217],
       [19.96174586],
       [24.1983249 ],
       [21.00202622],
       [29.38682967],
       [16.33215467],
       [19.37439646],
       [14

In [42]:
# finding the error as compared to the y dataset
print(mean_squared_error(test_y,pred_y))
#or
import numpy as np
np.mean((pred_y-test_y)**2)

19.831323672063235


MEDV    19.831324
dtype: float64

In [43]:
# finding the regression model for the original dataset
print(reg_model.score(train_x,train_y))

0.7103879080674731


### R-squared >=0.85 -BEST FIT
### R-squared>=0.7 -GOOD FIT
### R-squared>=0.5 -poor fit
 as the value of r here is .74 so this is good fit model

In [44]:
print(reg_model.intercept_) # prints the intercept of the value

[46.39649387]


In [45]:
print(reg_model.coef_) # print the coeefficients of the regression model

[[-9.85424717e-02  6.07841138e-02  5.91715401e-02  2.43955988e+00
  -2.14699650e+01  2.79581385e+00  3.57459778e-03 -1.51627218e+00
   3.07541745e-01 -1.12800166e-02 -1.00546640e+00  6.45018446e-03
  -5.68834539e-01]]
